security/pkg/k8s/chiron/utils.go

```diff
@@ -28,12 +28,194 @@ import (
 
 	"github.com/ghodss/yaml"
 
+	"istio.io/istio/pkg/spiffe"
+
+	"istio.io/istio/security/pkg/pki/util"
 	"istio.io/pkg/log"
 
+	cert "k8s.io/api/certificates/v1beta1"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// Generate a certificate and key from k8s CA
+// Working flow:
+// 1. Generate a CSR
+// 2. Submit a CSR
+// 3. Approve a CSR
+// 4. Read the signed certificate
+// 5. Clean up the artifacts (e.g., delete CSR)
+func genKeyCertK8sCA(wc *WebhookController, secretName string, secretNamespace, svcName string) ([]byte, []byte, []byte, error) {
+	// 1. Generate a CSR
+	// Construct the dns id from service name and name space.
+	// Example: istio-pilot.istio-system.svc, istio-pilot.istio-system
+	id := fmt.Sprintf("%s.%s.svc,%s.%s", svcName, secretNamespace, svcName, secretNamespace)
+	options := util.CertOptions{
+		Host:       id,
+		RSAKeySize: keySize,
+		IsDualUse:  false,
+		PKCS8Key:   false,
+	}
+	csrPEM, keyPEM, err := util.GenCSR(options)
+	if err != nil {
+		log.Errorf("CSR generation error (%v)", err)
+		return nil, nil, nil, err
+	}
+
+	// 2. Submit a CSR
+	csrName := fmt.Sprintf("domain-%s-ns-%s-secret-%s", spiffe.GetTrustDomain(), secretNamespace, secretName)
+	k8sCSR := &cert.CertificateSigningRequest{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "certificates.k8s.io/v1beta1",
+			Kind:       "CertificateSigningRequest",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: csrName,
+		},
+		Spec: cert.CertificateSigningRequestSpec{
+			Request: csrPEM,
+			Groups:  []string{"system:authenticated"},
+			Usages: []cert.KeyUsage{
+				cert.UsageDigitalSignature,
+				cert.UsageKeyEncipherment,
+				cert.UsageServerAuth,
+				cert.UsageClientAuth,
+			},
+		},
+	}
+	r, err := wc.certClient.CertificateSigningRequests().Create(k8sCSR)
+	if err != nil {
+		if !kerrors.IsAlreadyExists(err) {
+			log.Debugf("failed to create CSR (%v): %v", csrName, err)
+			return nil, nil, nil, err
+		}
+		//Otherwise, delete the existing CSR and create again
+		log.Debugf("delete an existing CSR: %v", csrName)
+		err = wc.certClient.CertificateSigningRequests().Delete(csrName, nil)
+		if err != nil {
+			log.Errorf("failed to delete CSR (%v): %v", csrName, err)
+			return nil, nil, nil, err
+		}
+		log.Debugf("create CSR (%v) after the existing one was deleted", csrName)
+		r, err = wc.certClient.CertificateSigningRequests().Create(k8sCSR)
+		if err != nil {
+			log.Debugf("failed to create CSR (%v): %v", csrName, err)
+			return nil, nil, nil, err
+		}
+	}
+	log.Debugf("CSR (%v) is created: %v", csrName, r)
+
+	// 3. Approve a CSR
+	log.Debugf("approve CSR (%v) ...", csrName)
+	r.Status.Conditions = append(r.Status.Conditions, cert.CertificateSigningRequestCondition{
+		Type:    cert.CertificateApproved,
+		Reason:  "k8s CSR is approved",
```

Comment:
Can you make the reason specific: "CSR for the webhook certificate xxx". Will this facilitate the debugging from apiserver log?

Reply:
Done.